### ISSUE TITLE
Does not handle fragmented packets properly.

### ISSUE THREAD

**Issue by sduensin** [signal11/hidapi#164]
From:  http://www.cs.unc.edu/Research/vrpn/troubleshooting.html

--- Cut ---

USB HID device not working under Linux

We had this problem for the Xkeys device. The max packet size of the HID endpoints was causing us problems. The reports were longer than the max packet size, so got fragmented across two packets. The libusb implementation of HIDAPI doesn't really handle this correctly. If you find another device that gives trouble, please email the list server to let us know. We'll either fix it, or start down the path of one of the more general fixes below.

Alan Ott says: If you want to, try adjusting the length in hid-libusb.c line 658 (const size_t length). Right now, it's set to the endpoint max packet length. The way USB works, it treats every packet which is the same length as the endpoint max packet length to mean that there is another packet (continuation) which is coming. It treats them all this way until there is a packet shorter than the max packet length. A short packet indicates the end of transfer. A zero-length packet needs to be sent if the transfer length is equal to the max packet length.

On top of that, libusb will buffer them up until there is either a short packet or the buffer is filled up (which can be adjusted by the size_t length variable I mentioned above).

If you want to try another work-around, the hidraw implementation (change the include to use linux/hid.c instead of linux/hid-libusb.c) should handle this correctly, but read the README about using the hidraw implementation. We had trouble using hidraw when we switched VRPN from native implementations to HIDAPI.

--- Cut ---

I tried adjusting the length variable.  No joy.  Setting it to 512 (maximum HID packet size) causes HIDAPI to stop returning packets.

This seems like a rather big problem.  I'm not even working with vrpn.  Just happened to see they have the same issue.

**Comment by karlp** [comment]
yes please :)  Doing app side hacks based on vid/pid at a later date is non-ideal  Further, should be just a duplicate of https://github.com/signal11/hidapi/issues/99 ? Do we need them both open?

**Comment by signal11** [comment]
> Doing app side hacks based on vid/pid at a later date is non-ideal

It's not app-side hacks. The protocol is what's agreed to between your host and device. True, it's described by the report descriptor, but if you're using HIDAPI to communicate with it (by sending raw report bytes), then you _know_ what the protocol is. The VID/PID indicates the device. Ask for the proper number of bytes based on what you know the device is (because of the VID/PID).[1]

Sure, it would be nice to parse the report descriptor and get all of this. It's a fair amount of work to get all the way correct.


[1] If you have multiple incompatible device types in the wild with the same VID/PID and version, then you've made yourself a mess, and you might be right to consider it an "app-side hack."

**Comment by igok99** [comment]
> Setting the length to 512, or some arbitrarily large number, will cause libusb to buffer transactions until it gets to that number of bytes (or until it receives a short transaction), and then return all the data as a single transfer. This is probably not what you want.

If the host configured to read transactions of 1024 and the device sends a last transaction that is less than 1024, how the host knows it is a short transaction? does this transaction should contain a special flag? 

thanks.
